services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -0,0 +1,215 @@
+package cloudfunctions
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"google.golang.org/api/cloudfunctions/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/idtoken"
+
+	"google.golang.org/api/option"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/services/streammanager/common"
+	"github.com/tidwall/gjson"
+)
+
+type Config struct {
+	Credentials            string `json:"credentials"`
+	FunctionEnvironment    string `json:"functionEnvironment"`
+	RequireAuthentication  bool   `json:"requireAuthentication"`
+	GoogleCloudFunctionUrl string `json:"googleCloudFunctionUrl"`
+}
+
+type Client struct {
+	service *cloudfunctions.Service
+	opts    common.Opts
+}
+
+var pkgLogger logger.Logger
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("streammanager").Child("GoogleCloudFunction")
+}
+
+type GoogleCloudFunctionProducer struct {
+	client *Client
+}
+
+// NewProducer creates a producer based on destination config
+func NewProducer(destination *backendconfig.DestinationT, o common.Opts) (*GoogleCloudFunctionProducer, error) {
+	var config Config
+	jsonConfig, err := json.Marshal(destination.Config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] Error while marshalling destination config :: %w", err)
+	}
+	err = json.Unmarshal(jsonConfig, &config)
+	if err != nil {
+		return nil, fmt.Errorf("[GoogleCloudFunction] error  :: error in GoogleCloudFunction while unmarshalling destination config:: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(config.Credentials)),
+	}
+
+	service, err := generateService(destination, opts...)
+	// If err is not nil then retrun
+	if err != nil {
+		pkgLogger.Errorf("[GoogleCloudFunction] error  :: %w", err)
+		return nil, err
+	}
+	client := &Client{service, o}
+
+	return &GoogleCloudFunctionProducer{client}, err
+}
+
+func (producer *GoogleCloudFunctionProducer) Produce(jsonData json.RawMessage, destConfig Config) (statusCode int, respStatus, responseMessage string) {
+
+	parsedJSON := gjson.ParseBytes(jsonData)
+	if destConfig.FunctionEnvironment == "gen1" && destConfig.RequireAuthentication {
+		client := producer.client
+
+		if client == nil {
+			respStatus = "Failure"
+			responseMessage = "[GoogleCloudFunction] error  :: Failed to initialize GoogleCloudFunction client"
+			return 400, respStatus, responseMessage
+		}
+
+		functionName := getFunctionName(destConfig.GoogleCloudFunctionUrl)
```

Comment:
can we move this to initialisation phase as this need not be done for every event, in the init phase you can even call google api to return proper value, instead of custom parsing logic

Reply:
Actually, I wasn't able to find any API to get the functionName from the URL, that's why I added some custom logic. Also, this will be needed only in case of `gen1` function not for all.